### PR TITLE
#930 add idempotency key field

### PR DESCRIPTION
**Pull request by lojzatran**
Related doc from adyen: https://docs.adyen.com/development-resources/api-idempotency

Boy-scout change: I also tried to fix one frequently failing test: https://github.com/commercetools/commercetools-adyen-integration/runs/7915704142?check_suite_focus=true#step:11:247

Addressing the issue: https://github.com/commercetools/commercetools-adyen-integration/issues/930

### CODE REVIEW

File: docs/adr/0008-idempotency-key.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# 7. Idempotency key
```
Comment (lojzatran) [inline docs/adr/0008-idempotency-key.md:1]:
```suggestion
# 8. Idempotency key
```

File: docs/adr/0008-idempotency-key.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# 8. Idempotency key
+
+Date: 2022-08-17
+
+## Status
+
+Approved
```
Comment (lojzatran) [inline docs/adr/0008-idempotency-key.md:7]:
```suggestion
Approved (https://github.com/commercetools/commercetools-adyen-integration/pull/1000)
```

File: docs/adr/0008-idempotency-key.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# 8. Idempotency key
+
+Date: 2022-08-17
+
+## Status
+
+Approved (https://github.com/commercetools/commercetools-adyen-integration/pull/1000)
```
Comment (lojzatran) [inline docs/adr/0008-idempotency-key.md:7]:
```suggestion
[Approved](https://github.com/commercetools/commercetools-adyen-integration/pull/1000)
```

File: docs/adr/0008-idempotency-key.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+To enable retry on failure for payment modifications, an idempotency key needs to be sent with every request. This idempotency key is sent in the header as `Idempotency-Key:<key>`. It has to be unique per request so that we can retry the request with the same idempotency key.
+
+There are 2 payment modification actions currently supported in Adyen-integration and that could be affected by the idempotency problem: [partial capture](../../extension/docs/ManualCapture.md) and [partial refund](../../extension/docs/Refund.md). Both payment modifications are triggered by adding a payment transaction.   
+
+## Decision
+A new String custom field called `idempotencyKey` will be added to the payment transaction. When this field is present, Adyen-integration will take its value and send it as as a value for `Idempotency-Key` header.
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply docs/adr/0008-idempotency-key.md:17]: Can you exactly write where should I change it using Github's Add a suggestion feature?

[the remark this answers is not in the document]
Comment (praveenkumarct) [reply docs/adr/0008-idempotency-key.md:17]: I think the below sentence would be good enough, 

`A new String custom field called `idempotencyKey` will be added to the payment transaction custom type with key `ctp-adyen-integration-transaction-payment-type`.`

File: extension/src/paymentHandler/make-payment.handler.js [hunk 5]
```diff
@@ -32,7 +32,8 @@ async function execute(paymentObject) {
     ),
   ]
 
-  const paymentMethod = request.paymentMethod?.type
+  const requestBodyJson = JSON.parse(request.body)
+  const paymentMethod = requestBodyJson.paymentMethod?.type
```
Comment (praveenkumarct) [inline extension/src/paymentHandler/make-payment.handler.js:36]:
```suggestion
  const paymentMethod = requestBodyJson?.paymentMethod?.type
```

File: notification/.prettierignore [hunk 6]
```diff
@@ -5,3 +5,4 @@ package.json
 package-lock.json
 .prettierignore
 .eslintrc
+.nyc_output
```
Comment (praveenkumarct) [inline notification/.prettierignore:8]: Same here, We should delete this file.

**Review by praveenkumarct — CHANGES_REQUESTED** [review]

File: docs/adr/0008-idempotency-key.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+To enable retry on failure for payment modifications, an idempotency key needs to be sent with every request. This idempotency key is sent in the header as `Idempotency-Key:<key>`. It has to be unique per request so that we can retry the request with the same idempotency key.
+
+There are 2 payment modification actions currently supported in Adyen-integration and that could be affected by the idempotency problem: [partial capture](../../extension/docs/ManualCapture.md) and [partial refund](../../extension/docs/Refund.md). Both payment modifications are triggered by adding a payment transaction.   
+
+## Decision
+A new String custom field called `idempotencyKey` will be added to the payment transaction. When this field is present, Adyen-integration will take its value and send it as a value for `Idempotency-Key` header.
```
Comment (lojzatran) [inline docs/adr/0008-idempotency-key.md:17]:
```suggestion
A new String custom field called `idempotencyKey` will be added to the payment transaction custom type `key="ctp-adyen-integration-transaction-payment-type"`. When this field `idempotencyKey` is present in the payment transaction, Adyen-integration will take its value and send it as a value for `Idempotency-Key` header. 
```

File: docs/adr/0008-idempotency-key.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply docs/adr/0008-idempotency-key.md]: https://github.com/commercetools/commercetools-adyen-integration/pull/1000/commits/8a0e8414452a0b10e7c9ca06f3f304cd94d07a8b

File: extension/.prettierignore [hunk 8]
```diff
@@ -5,3 +5,5 @@ package.json
 package-lock.json
 .prettierignore
 .eslintrc
+.nyc_output
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply extension/.prettierignore]: https://github.com/commercetools/commercetools-adyen-integration/pull/1000/commits/d950d2b667b96ecdc7fbad0c060880aae3255333

File: notification/.prettierignore (hunk 6 above)
Comment (lojzatran) [reply notification/.prettierignore]: https://github.com/commercetools/commercetools-adyen-integration/pull/1000/commits/b0b1d5cb42203115281398c7718a5ab3e03accc4

**Review by praveenkumarct — APPROVED** [review]

**Pull request merged**